magento/magento2
# [2.3] Reworked gallery.phtml to move generation of gallery json strings to own block functions

### Description
Reworked gallery.phtml to move generation of gallery json strings to own block functions
Introduced unit tests to cover generation of gallery json strings with full type checking.

The resultant code is much easier to diagnose and test. It uses the json_encode and json_decode routines instead of the old manual way of generating json strings that were used in the old gallery.phtml which is prone to error and mistakes.

### Fixed Issues (if relevant)
1. none. Rework and unit tests.

### Manual testing scenarios
1. View source on any product page.
2. Ensure the gallery section has fully formed json for gallery and gallery fullscreen sections.

Stock Magento should be

```
"options": {"nav":"thumbs","loop":true,"keyboard":true,"arrows":true,"allowfullscreen":true,"showCaption":false,"width":700,"thumbwidth":88,"thumbheight":110,"height":700,"transitionduration":500,"transition":"slide","navarrows":true,"navtype":"slides","navdir":"horizontal"}
```
and
```
"fullscreen": "nav":"thumbs","loop":true,"navdir":"horizontal","navarrows":false,"navtype":"slides","arrows":true,"showCaption":false,"transitionduration":500,"transition":"slide"}
```

All strings/bools/ints are now typed properly as per the Magento dev docs for the [Magento 2 Gallery Widget](https://devdocs.magento.com/guides/v2.3/javascript-dev-guide/widgets/widget_gallery.html).

### Contribution checklist
